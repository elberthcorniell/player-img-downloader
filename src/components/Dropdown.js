import React, { useState } from 'react';
import PropTypes from 'prop-types';
import {
    Button,
    Dropdown,
    FormControl,
} from 'react-bootstrap';

const DropdownButton = React.forwardRef(({ children, onClick }, ref) => (
    <Button
        ref={ref}
        style={{
            width: '100%',
            marginBottom: 10,
        }}
        onClick={onClick}>
        {children} &#x25bc;
    </Button>
));

const DropdownMenu = React.forwardRef(
    ({ children, style, className, 'aria-labelledby': labeledBy }, ref) => {
        const [value, setValue] = useState('');

        return (
            <div ref={ref}
                style={style}
                className={className}
                aria-labelledby={labeledBy}
            >
                <FormControl
                    autoFocus
                    placeholder="Search"
                    className="mx-3 my-2 w-auto"
                    onChange={({ target: { value } }) => setValue(value)}
                    value={value}
                />
                <ul className="list-unstyled">
                    {React.Children.toArray(children).filter(child => {
                        const search = value.toLowerCase();
                        const elementValue = child.props.children.toLowerCase();
                        return !value || elementValue.includes(search);
                    })}
                </ul>
            </div>
        );
    },
);

const DropdownSearch = ({ buttonText, items, disabled, onClick }) => (
    <Dropdown>
        <Dropdown.Toggle
            as={DropdownButton}
            id="dropdown-custom-components">
            {buttonText}
        </Dropdown.Toggle>
        <Dropdown.Menu as={DropdownMenu} disabled={disabled}>
            {items?.map((item, index) =>
                <Dropdown.Item
                    key={item.name}
                    onClick={() => onClick(item)}
                    disabled={disabled}
                    eventKey={`${index + 1}`}>
                    {item.name}
                </Dropdown.Item>
            )}
        </Dropdown.Menu>
    </Dropdown>
);

DropdownSearch.propTypes = {
    buttonText: PropTypes.string,
    items: PropTypes.arrayOf(PropTypes.object),
    onClick: PropTypes.func,
    disabled: PropTypes.bool,
}

DropdownSearch.defaultProps = {
    buttonText: 'Select item',
    items: [],
    onClick: () => { },
    disabled: false,
}

export default DropdownSearch;
