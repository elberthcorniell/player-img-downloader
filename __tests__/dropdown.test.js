import { render, screen } from '@testing-library/react';
import '@testing-library/jest-dom/extend-expect';
import Dropdown from '../src/components/Dropdown';

describe('Dropdown Button', () => {
    test('Render as expected', () => {
        render(<Dropdown />)
        expect(document.body.innerHTML).toMatchSnapshot();
    });

    test('Render as expected', () => {
        render(<Dropdown
            items={[
                { name: 'test1' },
                { name: 'test2' },
            ]}
        ></Dropdown>)
        document.getElementsByTagName('button')[0].click();
        const textElement1 = screen.getAllByText('test1');
        const textElement2 = screen.getAllByText('test2');
        expect(typeof textElement1[0]).not.toBe('undefined');
        expect(typeof textElement2[0]).not.toBe('undefined');
    });
});