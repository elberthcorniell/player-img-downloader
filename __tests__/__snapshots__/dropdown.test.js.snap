// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Dropdown Button Render as expected 1`] = `"<div><div class=\\"dropdown\\"><button style=\\"width: 100%; margin-bottom: 10px;\\" type=\\"button\\" class=\\"btn btn-primary\\">Select item â–¼</button></div></div>"`;
